spatial/curve/hilbert.go

```diff
@@ -0,0 +1,254 @@
+// (https://en.wikipedia.org/w/index.php?title=Hilbert_curve&oldid=1011599190).
+// The implementation of Hilbert3D and Hilbert4D are extrapolated from
+// Hilbert2D.
+//
+// For the first-order k-dimensional Hilbert curve, a spatial point V is mapped
+// to a point on the curve D by XOR - each dimension of V is expected to be 0 or
+// 1:
+//
+//     func map1stOrder(k int, v []int) (d int) {
+//         for i := k - 1; i >= 0; i-- {
+//             d = d<<1 | (d^v[i])&1
+//         }
+//         return d
+//     }
+//
+// In a 2-space with the origin at the bottom left, this results in a ⊐ shape,
+// wound counter clockwise.
+package curve
+
+// NewHilbert returns a 2-, 3-, or 4-dimensional Hilbert curve. NewHilbert will
+// panic if dimension is not in the range [2, 4] or if order is not in the range
+// [1, ∞).
+//
+// The runtime of Space and Curve scales as O(k∙n) where k is the dimension and
+// n is the order. The length of the curve is 2^(k∙n).
+func NewHilbert(order, dimension int) SpaceFilling {
+	if order < 1 {
+		panic("order must be positive")
+	}
+
+	switch dimension {
+	case 2:
+		return Hilbert2D{Order: order}
+	case 3:
+		return Hilbert3D{Order: order}
+	case 4:
+		return Hilbert4D{Order: order}
+	default:
+		if dimension < 2 {
+			panic("dimensions must be 2 or greater")
+		}
+		panic("more than 4 dimensions are not supported")
+	}
+}
+
+// V may get mangled by Curve, so copy it
+func dup(v []int) []int {
+	u := make([]int, len(v))
+	copy(u, v)
+	return u
+}
+
+type coordInvert struct{ I, J int } // invert I and J
+type coordSwap struct{ I, J int }   // swap I and J
+type coordFlip struct{ I, J int }   // swap and invert I and J
+
+func (c coordInvert) do(n int, v []int) { v[c.I], v[c.J] = v[c.I]^(1<<n-1), v[c.J]^(1<<n-1) }
+func (c coordSwap) do(n int, v []int)   { v[c.I], v[c.J] = v[c.J], v[c.I] }
+func (c coordFlip) do(n int, v []int)   { v[c.I], v[c.J] = v[c.J]^(1<<n-1), v[c.I]^(1<<n-1) }
+
+// compose multiple coordinate modifications
+type multiCoord []interface{ do(int, []int) }
+
+func (c multiCoord) do(reverse bool, n int, v []int) {
+	if reverse {
+		for i := len(c) - 1; i >= 0; i-- {
+			c[i].do(n, v)
+		}
+	} else {
+		for _, c := range c {
+			c.do(n, v)
+		}
+	}
+}
+
+// Hilbert2D is a 2-dimensional Hilbert curve.
+type Hilbert2D struct{ Order int }
+
+// Size returns {2ⁿ, 2ⁿ} where n is the order.
+func (h Hilbert2D) Size() []int { return []int{1 << h.Order, 1 << h.Order} }
+
+func (h Hilbert2D) rot(n int, v []int, d int) {
+	switch d {
+	case 0:
+		coordSwap{0, 1}.do(n, v)
+	case 3:
+		coordFlip{0, 1}.do(n, v)
+	}
+}
+
+// Curve returns the curve coordinate of V.
+func (h Hilbert2D) Curve(v ...int) Point {
+	v = dup(v)
+	var d Point
+	for n := h.Order - 1; n >= 0; n-- {
+		rx := (v[0] >> n) & 1
+		ry := (v[1] >> n) & 1
+		rd := ry<<1 | (ry ^ rx)
+		d += Point(rd) << (2 * n)
+		h.rot(h.Order, v, rd)
+	}
+	return d
+}
+
+// Space2D returns the spatial coordinates of D.
+func (h Hilbert2D) Space2D(d Point) [2]int {
```

Comment:
Why do we have the array and slice versions of each of these?

Reply:
`Space(d Point) []int` allows all of the curves to share an interface, but it requires allocating a slice. `Space[N]D(d Point) [N]int` does not require an allocation, but it has a different signature for each type. So if you're using `Hilbert3D` directly, you can use the allocation-free version. But if you need to support different dimensionalities, you can use the one that returns a slice.

Reply:
That's what I was thinking, thanks.